algorithms/HPdclassifier/R/hpdrpart.R

```diff
@@ -0,0 +1,279 @@
+	free_sh_mem = free_sh_mem - max_nodes_per_iteration*sizeof_node_histogram
+
+	#use 80% of remaining memory to transfer data during local tree building step
+	nodes_per_executor= 1
+	threshold = as.integer(min(
+		  .8*free_mem/nodes_per_executor/ncol(observations),
+		  .8*free_sh_mem/nodes_per_executor/ncol(observations)))
+
+	if(threshold > nrow(observations))
+	{
+		threshold = nrow(observations)
+		nodes_per_executor = as.integer(min(
+				   free_mem/threshold/ncol(observations),
+				   free_sh_mem/threshold/ncol(observations)))
+	}
+
+	free_mem = free_mem - ncol(observations)*threshold*nodes_per_executor
+	free_sh_mem = free_sh_mem - ncol(observations)*threshold*nodes_per_executor
+
+
+	threshold = as.integer(floor(threshold))
+	max_nodes_per_iteration = as.integer(floor(max_nodes_per_iteration))
+	nodes_per_executor = as.integer(floor(nodes_per_executor))
+
+	threshold = min(threshold,nrow(observations)/sum(DR_status$Inst))
+
+	if(do.trace)
+		print(paste("threshold",
+			toString(threshold),sep=" = "))
+
+	if(do.trace)
+		print(paste("nodes_per_executor",
+			toString(nodes_per_executor),sep=" = "))
+
+	if(do.trace)
+		print(paste("max_nodes_per_iteration",
+			toString(max_nodes_per_iteration),sep=" = "))
+
+
+	suppressWarnings({
+	tree <- .hpdRF_distributed(observations, responses, ntree = 1L, 
+	     bin_max = as.integer(nBins), 
+	     features_cardinality = features_cardinality,
+	     response_cardinality = response_cardinality, 
+	     features_num = as.integer(ncol(observations)),
+	     threshold = threshold, weights = weights, 
+	     nodes_per_worker=nodes_per_executor,
+	     max_nodes = .Machine$integer.max, node_size = control$minsplit, 
+	     replacement = TRUE, cutoff = cutoff, classes = classes, 
+	     completeModel = FALSE, 
+	     max_nodes_per_iteration = max_nodes_per_iteration,
+	     trace = do.trace, features_min = NULL, features_max = NULL,
+	     min_split = control$minbucket, max_depth = control$maxdepth, 
+	     cp = control$cp)
+	})
+	     
+	if(do.trace)
+	print("converting to rpart model")
+	timing_info <- Sys.time()
+	model = convertToRpartModel(tree$forest, x_colnames)
+	timing_info <- Sys.time() - timing_info
+	if(do.trace )
+	print(timing_info)
+
+
+	model$call = match.call()
+	model$terms = variables$terms
+	if(is.na(response_cardinality))
+		model$method = "regression"
+	if(!is.na(response_cardinality))
+		model$method = "classification"
+	model$control = control
+	model$params = params
+	model$na.action = na.action
+	model$numresp = 0
+	model$numresp = length(classes)
+	model$classes = classes
+	class(model) <- c("hpdrpart","rpart")
+
+	if(is.data.frame(data))
+		responses = getpartition(responses)
+
+	if(do.trace)
+	print("calculating variable importance")
+	timing_info <- Sys.time()
+	model$variable.importance <- 
+		varImportance(model,data,responses)
+	timing_info <- Sys.time() - timing_info
+	if(do.trace )
+	print(timing_info)
+
+	return(model)
+}
+
```

Comment:
Please add a deploy.hpdrpart function, and validate its input model before passing it to its output.